src/librustc_borrowck/borrowck/mir/dataflow/mod.rs

```diff
@@ -0,0 +1,1168 @@
+            |path, s| Self::update_bits(sets, path, s)
+        )
+    }
+
+    fn propagate_call_return(&self,
+                             ctxt: &Self::Ctxt,
+                             in_out: &mut [usize],
+                             _call_bb: repr::BasicBlock,
+                             _dest_bb: repr::BasicBlock,
+                             dest_lval: &repr::Lvalue) {
+        // when a call returns successfully, that means we need to set
+        // the bits for that dest_lval to 1 (initialized).
+        let move_path_index = ctxt.2.rev_lookup.find(dest_lval);
+        super::on_all_children_bits(
+            ctxt.0, ctxt.1, &ctxt.2,
+            move_path_index,
+            |mpi| { in_out.set_bit(mpi.idx()); }
+        );
+    }
+}
+
+fn zero_to_one(bitvec: &mut [usize], move_index: MoveOutIndex) {
+    let retval = bitvec.set_bit(move_index.idx());
+    assert!(retval);
+}
+
+impl<'a, 'tcx> BitwiseOperator for MovingOutStatements<'a, 'tcx> {
+    #[inline]
+    fn join(&self, pred1: usize, pred2: usize) -> usize {
+        pred1 | pred2 // moves from both preds are in scope
+    }
+}
+
+impl<'a, 'tcx> BitwiseOperator for MaybeInitializedLvals<'a, 'tcx> {
+    #[inline]
+    fn join(&self, pred1: usize, pred2: usize) -> usize {
+        pred1 | pred2 // "maybe" means we union effects of both preds
+    }
+}
+
+impl<'a, 'tcx> BitwiseOperator for MaybeUninitializedLvals<'a, 'tcx> {
+    #[inline]
+    fn join(&self, pred1: usize, pred2: usize) -> usize {
+        pred1 | pred2 // "maybe" means we union effects of both preds
+    }
+}
+
+impl<'a, 'tcx> BitwiseOperator for DefinitelyInitializedLvals<'a, 'tcx> {
+    #[inline]
+    fn join(&self, pred1: usize, pred2: usize) -> usize {
+        pred1 & pred2 // "definitely" means we intersect effects of both preds
+    }
+}
+
+// FIXME: `DataflowOperator::initial_value` should be named
+// `bottom_value`. The way that dataflow fixed point iteration works,
+// you want to start at bottom and work your way to a fixed point.
+// This needs to include the detail that the control-flow merges will
+// apply the `join` operator above to current state (which starts at
+// that bottom value).
+//
+// This means, for propagation across the graph, that you either want
+// to start at all-zeroes and then use Union as your merge when
+// propagating, or you start at all-ones and then use Intersect as
+// your merge when propagating.
+
+impl<'a, 'tcx> DataflowOperator for MovingOutStatements<'a, 'tcx> {
+    #[inline]
+    fn initial_value() -> bool {
+        false // bottom = no loans in scope by default
+    }
+}
+
+impl<'a, 'tcx> DataflowOperator for MaybeInitializedLvals<'a, 'tcx> {
+    #[inline]
+    fn initial_value() -> bool {
+        false // bottom = uninitialized
+    }
+}
+
+impl<'a, 'tcx> DataflowOperator for MaybeUninitializedLvals<'a, 'tcx> {
+    #[inline]
+    fn initial_value() -> bool {
+        false // bottom = uninitialized
```

Comment:
I think this should be `bottom = initialized`, since a `1` means "maybe uninit"?